keps/sig-api-machinery/20190325-unions.md

```diff
@@ -0,0 +1,213 @@
+have a keyword to implement “oneof” or "union". They allow APIs to have a better
+semantic, typically as a way to say “only one of the given fields can be
+set”. We currently have multiple occurrences of this semantic in kubernetes core
+types, at least:
+- VolumeSource is a structure that holds the definition of all the possible
+  volume types, only one of them must be set, it doesn't have a discriminator.
+- DeploymentStrategy is a structure that has a discrminator
+  "DeploymentStrategyType" which decides if "RollingUpate" should be set
+
+The problem with the lack of solution is that:
+- The API is implicit, and people don't know how to use it
+- Clients can't know how to deal with that, especially if they can't parse the
+  OpenAPI
+- Server can't understand the user intent and normalize the object properly
+
+## Motivation
+
+Currently, changing a value in an one-of type is difficult because the semantic
+is implicit, which means that nothing can be built to automatically fix unions,
+leading to many bugs and issues:
+- https://github.com/kubernetes/kubernetes/issues/35345
+- https://github.com/kubernetes/kubernetes/issues/24238
+- https://github.com/kubernetes/kubernetes/issues/34292
+- https://github.com/kubernetes/kubernetes/issues/6979
+- https://github.com/kubernetes/kubernetes/issues/33766
+- https://github.com/kubernetes/kubernetes/issues/24198
+- https://github.com/kubernetes/kubernetes/issues/60340
+
+And then, for other people:
+- https://github.com/rancher/rancher/issues/13584
+- https://github.com/helm/charts/pull/12319
+- https://github.com/EnMasseProject/enmasse/pull/1974
+- https://github.com/helm/charts/pull/11546
+- https://github.com/kubernetes/kubernetes/pull/35343
+
+This is replacing a lot of previous work and long-standing effort:
+- Initially: https://github.com/kubernetes/community/issues/229, then
+- https://github.com/kubernetes/community/pull/278
+- https://github.com/kubernetes/community/pull/620
+- https://github.com/kubernetes/kubernetes/pull/44597
+- https://github.com/kubernetes/kubernetes/pull/50296
+- https://github.com/kubernetes/kubernetes/pull/70436
+
+Server-side [apply](http://features.k8s.io/555) is what enables this proposal to
+become possible.
+
+### Goals
+
+The goal is to enable a union or "oneof" semantic in Kubernetes types, both for
+in-tree types and for CRDs.
+
+### Non-Goals
+
+We're not planning to use this KEP to release the feature, but mostly as a way
+to document what we're doing.
+
+## Proposal
+
+In order to support unions in a backward compatible way in kubernetes, we're
+proposing the following changes.
+
+Note that this proposes unions to be "at most one of". Whether exactly one is
+supported or not should be implemented by the validation logic.
+
+### Go tags
+
+We're proposing a new type of tags for go types (in-tree types, and also
+kubebuilder types):
+
+- `// +union` before a field means that it is part of the union. Multiple fields
```

Comment:
I think we should put this on types:

```go
// +union
type MetricSource struct {
  // +discriminant
  Type MetricSourceType
  Object *ObjectMetricSource
  Pods *PodsMetricSource
  ...
}
```

We should be thinking of unions like enums in Rust, algebraic datatypes in Haskell,  discriminated unions in typescript, etc: as a type-level construct with multiple variants and a discriminator, at both the IDL-ish (Go) level and at an API level.  This is conceptually similar to what people are familiar with, and makes writing things out easier.

This should hold for most cases in Kubernetes.  The few where it doesn't we can either a) special-case, b) fix before they go alpha (*cough* webhookconfig *cough*), or c) fix at a type level and leave the API the same by making an external struct and embedding.

Reply:
I also prefer a formulation like this one, I think. Embedding can potentially be used to transition existing places where union and non-union fields are mixed.

Reply:
What about something in the middle: `+union-fields=Object,Pods` on types.

Reply:
That seems more error-prone and tedious and hard to read? :-/

I'm going with both: you can either specify tags on members individually or entire structs. They'll generate the same OpenAPI, and we'll only document the struct way, which is the one we want people to use (and also is probably going to be the only one available through kubebuilder)

Reply:
> 'm going with both:

sounds good.